### PR TITLE
Hats example app

### PR DESCRIPTION
**Pull request by amberroy**
Example of synced objects that follow the user via the altspace avatar tracking skeleton API.  An initial set of stationary hats, one per type, appear floating in the enclosure when the app loads.  When a user clicks on a had, it is cloned and placed on top of their head.  The hat is destroyed when the user clicks on another hat, or leaves the room.

### CODE REVIEW

File: examples/hats.html [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	var scale;
+
+	function main() {
+
+		sim = altspace.utilities.Simulation();
+		var promises = [altspace.getThreeJSTrackingSkeleton(), altspace.getEnclosure()];
+		Promise.all(promises).then(function(array) {
+			skeleton = array[0];
+			head = skeleton.getJoint('Head');//head is an Object3D
+			if (!head) throw Error('Failed to get head joint');
```
Comment (GavanWilhite) [inline examples/hats.html]: will this happen?

**Comment by amberroy** [comment]
I'm going to dig into why joint.add() didn't work, may be due to forgetting to add the skeleton to the scene.

File: examples/hats.html [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+			obj.translateY(offset.y);
+			obj.translateZ(offset.z);
+			obj.addBehavior(altspace.utilities.behaviors.Object3DSync({
+				position: true, rotation: true, auto: true
+			}));
+		}
+		return obj;
+	}
+
+	function FollowUserBehavior(config){
```
Comment (GavanWilhite) [inline examples/hats.html]: can pull this now then?

File: examples/hats.html [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	function addEventListener(obj) {
+		obj.addEventListener('cursordown', function(event) {
+			if (selectedHat) sceneSync.destroy(selectedHat);
+			//destroyOnDisconnect=true so when user leaves, their hat vanishes.
+			var myHat = sceneSync.instantiate('Hat', {name: obj.name, isOnHead: true}, true);
+			//Add follow behavior here, not in createHat, since only the local hat
+			//follows this user (although its position is synced to all users).
+			myHat.addBehavior(FollowUserBehavior({scene: sim.scene}));
```
Comment (GavanWilhite) [inline examples/hats.html]: this is where we should add it to the trackingskeleton head right? (I think you might have doing that before) If it's in the instantiator then they will all be on the local user's head :P

File: examples/hats.html [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+		multiloader.load(loadRequest, onLoaded);
+	}
+
+	function onLoaded() {
+
+		for (var i=0; i < loadRequest.objects.length; i++) {
+			var objTemplate = loadRequest.objects[i];
+			var name = config.models[i].name;//guaranteed to be in same order
+			if (!objTemplate) throw Error('Error loading hat model ' + name);
+			var v3 = new THREE.Vector3();
```
Comment (GavanWilhite) [inline examples/hats.html]: can this be called 'offset'?

**Comment by amberroy** [comment]
Ready for review.

File: examples/hats.html [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		sim = altspace.utilities.Simulation();
+		var promises = [altspace.getThreeJSTrackingSkeleton(), altspace.getEnclosure()];
+		Promise.all(promises).then(function(array) {
+			skeleton = array[0];
+			sim.scene.add(skeleton);
+			head = skeleton.getJoint('Head');//head is an Object3D
+			if (!head) throw Error('Failed to get head joint');
+			enclosure = array[1];
+		  scale = (1.0 * config.modelsPixelsPerMeter) / enclosure.pixelsPerMeter;
+		  loadModels();
```
Comment (brianpeiris) [inline examples/hats.html]: Minor comment: The indentation is off here. Use tabs.

File: examples/hats.html [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+	function main() {
+
+		sim = altspace.utilities.Simulation();
+		var promises = [altspace.getThreeJSTrackingSkeleton(), altspace.getEnclosure()];
+		Promise.all(promises).then(function(array) {
+			skeleton = array[0];
+			sim.scene.add(skeleton);
+			head = skeleton.getJoint('Head');//head is an Object3D
+			if (!head) throw Error('Failed to get head joint');
```
Comment (brianpeiris) [inline examples/hats.html]: If this is an example for users to follow, this line of code implies that the skeleton doesn't always have a head. Is that true?

Comment (amberroy) [inline examples/hats.html]: True this could be confusing, removed that line.

File: examples/hats.html (hunk 5 above)
Comment (amberroy) [inline examples/hats.html]: opps thanks.

**Pull request merged**